Makefile

```diff
@@ -108,14 +109,33 @@ test/gc/node_modules/weak/build/Release/weakref.node: $(NODE_EXE)
 		--directory="$(shell pwd)/test/gc/node_modules/weak" \
 		--nodedir="$(shell pwd)"
 
-build-addons: $(NODE_EXE)
-	rm -rf test/addons/doc-*/
-	./$(NODE_EXE) tools/doc/addon-verify.js
-	$(foreach dir, \
-			$(sort $(dir $(wildcard test/addons/*/*.gyp))), \
-			./$(NODE_EXE) deps/npm/node_modules/node-gyp/bin/node-gyp rebuild \
-					--directory="$(shell pwd)/$(dir)" \
-					--nodedir="$(shell pwd)" && ) echo "build done"
+ADDONS_TESTS = \
+	async-hello-world \
+	at-exit \
+	hello-world-function-export \
+	hello-world \
+	repl-domain-abort \
+
+# Note: don't depend on $(NODE_EXE) here, that unconditionally forces a rebuild.
+test/addons/%/build/Release/binding.node: \
+		test/addons/%/binding.cc test/addons/%/binding.gyp
+	./$(NODE_EXE) deps/npm/node_modules/node-gyp/bin/node-gyp rebuild \
+		--nodedir="$(PWD)" --directory="$(dir $<)"
+
+build-static-addons: \
+	$(NODE_EXE) $(ADDONS_TESTS:%=test/addons/%/build/Release/binding.node)
+
+# Note: don't depend on $(NODE_EXE) here, that unconditionally forces a rebuild.
+test/addons/.stamp: tools/doc/addon-verify.js doc/api/addons.markdown
```

Comment:
Oh! That's the problem -- it evaluates and expands `$(foreach dir...` **before** executing `addon-verify.js` -- and thus it can't see those new directories, and the test fails because node-gyp hasn't run in those dirs.